### ISSUE TITLE
Special Token for color values possible?

### ISSUE THREAD

**Issue by void256** [phil-brown/jCSS-Parser#3]
For my project I need to parse the "linear-gradient" CSS rule.

Since jCSS-Parser does not support that directly I need to do that on my own. So I've build my own little parser using the jCSS-Parser classes `TokenSequence` and `Token` as its input. This works quite well and I can parse the "linear-gradient" rule pretty well :)

However, with color values like "#fff" or "#ffffff" I've encountered a problem. Since the 'Scanner' class returns the color values as a combination of Token.Number and Token.Identifier - depending on the first character being a number (0-9) or a hexadecimal number beginning with (a-f) - it's a bit difficult to process. Additionally CSS color values could consist of 3 digits or 6 digits too so it's quite hard to parse them correctly.

Especially difficult for the linear-gradient syntax since they can be followed directly by a color stop number too.

Here are some examples:

```
linear-gradient(#200b30 10%);
linear-gradient(#a0b29f 10%);
linear-gradient(#0aa 10%);
```

And the way they are separated into tokens by the Scanner:

```
"#200b30 10%"
102:null "#"
126:200 "200"
127:b30 "b30"
126:10 "10"
138:null "%"

"#a0b29f 10%"
102:null "#"
127:a0b29f "a0b29f"
126:10 "10"
138:null "%"

"#0aa 10%"
102:null "#"
126:0 "0"
127:aa "aa"
126:10 "10"
138:null "%"
```

The problem is that we get the color values returned as a combination of Token.Number and Token.Identifier. Since a color value could consist of 3 or 6 digits one could concat the Token.attribute values into a String and check it's length to know if the color value has been parsed completely but that is not quite possible in all cases.

In the last example above we get back "0" and "aa". So combined we get the color value "0aa" which is a 3 digit value BUT since it is followed by another number "10" we could "think" it might be a 6 digit color value and continue processing it as a color value into "0aa10" and then run into an error since the following "%" token is not expected in that case.

I hope I make any sense :wink: since it's kinda hard to explain :smile:

I think this approach could work if we would get back the whitespace "token" too so that we could find the border between the color value and the next stop number but as far as I see it's not returned as a Token.

I think it might be best if the Scanner would return a Token.Color value directly but I don't understand enough of the CSS spec and the Scanner to really be certain if that's allowed / makes sense at all.

Maybe I'm parsing the color values completely wrong and there is actually a way to read them smarter? In that case I'd be grateful for any hints too :smile:

**Comment by phil-brown** [comment]
@void256 This makes a lot of sense. I have a partial implementation of this more granular parser in my [droidQuery](https://github.com/phil-brown/droidQuery/tree/master/droidQuery/src/self/philbrown/css) project. I originally left out declaration value parsing from _jCSS-Parser_, so it could be platform independent (droidQuery is for Android, but I wanted to support [javaQuery](https://github.com/phil-brown/javaQuery) at some point too - plus be an option for others' projects). For your implementation, I would recommend something similar to [StyleSheet.java](https://github.com/phil-brown/droidQuery/blob/master/droidQuery/src/self/philbrown/css/StyleSheet.java), which uses _reflection_ to call methods named by a _Declaration_'s _property_, and passing in the _Declaration_'s _value_ as a _TokenSequence_, which you can then parse and handle how it works best in your project. My implementation does allow spaces and methods (could be used with _linear-gradient_, but currently I do not have support for that).

I will start looking into a more granular, platform-specific (Android and Java) solution for this when I can :smile:
